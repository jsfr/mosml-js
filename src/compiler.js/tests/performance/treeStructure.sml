prim_val ctimebegin_ : string -> unit = 1 "console.time";
prim_val ctimeend_ : string -> unit = 1 "console.timeEnd";
fun ctimeBegin n = ctimebegin_ n;
fun ctimeEnd n = ctimeend_ n;

fun repeat n f x =
    let val n = ref n
    in  while !n > 0 do (n := !n-1; f x) end;
fun timerep n f s = repeat n (fn x => (ctimeBegin(s);f x; ctimeEnd(s)));

datatype Tree =
    Node of int * Tree * Tree
  | Leaf

fun rev xs =
  let
    fun revAcc [] ys = ys
    | revAcc (x::xs) ys = revAcc xs (x::ys)
  in
    revAcc xs []
  end

fun intList 0 = [0]
  | intList n = n :: intList (n-1)

fun genTree [] = Leaf
  | genTree (x::xs) = Node (x, genTree xs, genTree (rev xs))

fun traverseTree (Leaf, _) = 0
  | traverseTree (Node(x, left, right), n) =
    if n = x
    then 1 + traverseTree (left, n) + traverseTree (right, n)
    else traverseTree (left, n) + traverseTree (right, n)

val _ = timerep 100 genTree "genTree" (intList 19)

val tree = genTree (intList 19)
val _ = timerep 100 traverseTree "traverseTree" (tree, 0)