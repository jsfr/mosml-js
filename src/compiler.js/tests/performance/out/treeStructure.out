beautified treeStructure.js
==============================JS BEGIN==============================
genTree: 2151ms
genTree: 2278ms
genTree: 2596ms
genTree: 2198ms
genTree: 2363ms
genTree: 2325ms
genTree: 2382ms
genTree: 2293ms
genTree: 2326ms
genTree: 2343ms
genTree: 2223ms
genTree: 2345ms
genTree: 2234ms
genTree: 2323ms
genTree: 2337ms
genTree: 2196ms
genTree: 2532ms
genTree: 2353ms
genTree: 2299ms
genTree: 2213ms
genTree: 2369ms
genTree: 2342ms
genTree: 2233ms
genTree: 2357ms
genTree: 2264ms
genTree: 2405ms
genTree: 2311ms
genTree: 2327ms
genTree: 2342ms
genTree: 2354ms
genTree: 2217ms
genTree: 2299ms
genTree: 2243ms
genTree: 2568ms
genTree: 2208ms
genTree: 2418ms
genTree: 2362ms
genTree: 2346ms
genTree: 2306ms
genTree: 2513ms
genTree: 2222ms
genTree: 2355ms
genTree: 2253ms
genTree: 2315ms
genTree: 2485ms
genTree: 2230ms
genTree: 2340ms
genTree: 2240ms
genTree: 2448ms
genTree: 2409ms
genTree: 2391ms
genTree: 2337ms
genTree: 2343ms
genTree: 2388ms
genTree: 2236ms
genTree: 2468ms
genTree: 2320ms
genTree: 2206ms
genTree: 2415ms
genTree: 2310ms
genTree: 2350ms
genTree: 2382ms
genTree: 2272ms
genTree: 2277ms
genTree: 2442ms
genTree: 2353ms
genTree: 2406ms
genTree: 2595ms
genTree: 2404ms
genTree: 2188ms
genTree: 2441ms
genTree: 2442ms
genTree: 2272ms
genTree: 2445ms
genTree: 2366ms
genTree: 2280ms
genTree: 2280ms
genTree: 2473ms
genTree: 2422ms
genTree: 2388ms
genTree: 2307ms
genTree: 2329ms
genTree: 2415ms
genTree: 2439ms
genTree: 2440ms
genTree: 2206ms
genTree: 2283ms
genTree: 2436ms
genTree: 2540ms
genTree: 2448ms
genTree: 2256ms
genTree: 2254ms
genTree: 2509ms
genTree: 2227ms
genTree: 2426ms
genTree: 2198ms
genTree: 2265ms
genTree: 2333ms
genTree: 2444ms
genTree: 2436ms
traverseTree: 745ms
traverseTree: 701ms
traverseTree: 738ms
traverseTree: 692ms
traverseTree: 675ms
traverseTree: 702ms
traverseTree: 676ms
traverseTree: 737ms
traverseTree: 685ms
traverseTree: 684ms
traverseTree: 693ms
traverseTree: 691ms
traverseTree: 687ms
traverseTree: 699ms
traverseTree: 674ms
traverseTree: 675ms
traverseTree: 684ms
traverseTree: 684ms
traverseTree: 667ms
traverseTree: 699ms
traverseTree: 676ms
traverseTree: 664ms
traverseTree: 839ms
traverseTree: 665ms
traverseTree: 665ms
traverseTree: 837ms
traverseTree: 671ms
traverseTree: 740ms
traverseTree: 698ms
traverseTree: 685ms
traverseTree: 763ms
traverseTree: 668ms
traverseTree: 671ms
traverseTree: 696ms
traverseTree: 666ms
traverseTree: 668ms
traverseTree: 812ms
traverseTree: 665ms
traverseTree: 668ms
traverseTree: 718ms
traverseTree: 667ms
traverseTree: 668ms
traverseTree: 787ms
traverseTree: 669ms
traverseTree: 665ms
traverseTree: 696ms
traverseTree: 665ms
traverseTree: 666ms
traverseTree: 707ms
traverseTree: 666ms
traverseTree: 688ms
traverseTree: 710ms
traverseTree: 678ms
traverseTree: 707ms
traverseTree: 691ms
traverseTree: 680ms
traverseTree: 686ms
traverseTree: 673ms
traverseTree: 693ms
traverseTree: 717ms
traverseTree: 686ms
traverseTree: 677ms
traverseTree: 690ms
traverseTree: 667ms
traverseTree: 671ms
traverseTree: 697ms
traverseTree: 668ms
traverseTree: 665ms
traverseTree: 764ms
traverseTree: 668ms
traverseTree: 676ms
traverseTree: 699ms
traverseTree: 680ms
traverseTree: 672ms
traverseTree: 778ms
traverseTree: 694ms
traverseTree: 680ms
traverseTree: 729ms
traverseTree: 664ms
traverseTree: 677ms
traverseTree: 691ms
traverseTree: 666ms
traverseTree: 820ms
traverseTree: 688ms
traverseTree: 668ms
traverseTree: 760ms
traverseTree: 665ms
traverseTree: 663ms
traverseTree: 702ms
traverseTree: 686ms
traverseTree: 666ms
traverseTree: 753ms
traverseTree: 673ms
traverseTree: 669ms
traverseTree: 695ms
traverseTree: 681ms
traverseTree: 667ms
traverseTree: 686ms
traverseTree: 666ms
traverseTree: 664ms
==============================JS END================================
